services/QuillLMS/app/services/diagnostic_insights/skill_practice_by_classroom.rb

```diff
@@ -0,0 +1,68 @@
+# frozen_string_literal: true
+
+module DiagnosticInsights
+  class SkillPracticeByClassroom < ApplicationService
+    class UnknownDiagnosticType < StandardError; end
+
+    attr_reader :classroom_id, :diagnostic_type, :start_date, :end_date
+
+    def initialize(classroom_id, diagnostic_type, start_date: nil, end_date: nil)
+      @classroom_id = classroom_id
+      @diagnostic_type = validate_diagnostic_type(diagnostic_type)
+      @start_date = start_date || Constants::DEFAULT_START_DATE
+      @end_date = end_date || Float::INFINITY
+    end
+
+    def run
+      students.map do |student|
+        {
+          user_id: student.id,
+          user_name: student.name,
+          skills: format_skills_for_student(student)
+        }
+      end
+    end
+
+    private def format_skills_for_student(student)
+      student_activity_sessions = activity_sessions_by_student[student.id]
+
+      return [] unless student_activity_sessions
+
+      student_activity_sessions
+        .group_by { |activity_session| activity_session.skill_performance.first['skill_name'] }
```

Comment:
Do we need safe navigation here?
`activity_session.skill_performance.first&.fetch('skill_name')`